packages/init/src/SliceMachineInitProcess.ts

```diff
@@ -0,0 +1,1114 @@
+					assertExists(
+						this.context.installProcess,
+						"Initial dependencies installation process must be available through context to run `finishCoreDependenciesInstallation`",
+					);
+
+					const updateOutput = (data: Buffer | null) => {
+						if (data instanceof Buffer) {
+							task.output = data.toString();
+						}
+					};
+
+					// Don't clutter console with logs when process is non TTY (CI, etc.)
+					if (process.stdout.isTTY || process.env.NODE_ENV === "test") {
+						this.context.installProcess.stdout?.on("data", updateOutput);
+					}
+					this.context.installProcess.stderr?.on("data", updateOutput);
+
+					try {
+						await this.context.installProcess;
+					} catch (error) {
+						/**
+						 * Error catching happens when then process is started ealier so
+						 * that all install errors, earlier and presents, can be catched.
+						 *
+						 * Here, we force the task to wait so that it is neither marked as
+						 * done or has the opportunity to handle the error itself.
+						 */
+						// If for whatever reason the process is not exited by now, we still throw the error
+						await new Promise((resolve) => setTimeout(resolve, 5000));
+						throw error;
+					}
+
+					task.title = `Installed core dependencies with ${chalk.cyan(
+						this.context.packageManager,
+					)}`;
+				},
+			},
+		]);
+	}
+
+	protected upsertSliceMachineConfigurationAndStartPluginRunner(): Promise<void> {
+		return listrRun([
+			{
+				title: "Resolving Slice Machine configuration...",
+				task: async (_, parentTask) => {
+					assertExists(
+						this.context.framework,
+						"Project framework must be available through context to run `upsertSliceMachineConfiguration`",
+					);
+					assertExists(
+						this.context.repository,
+						"Repository selection must be available through context to run `upsertSliceMachineConfiguration`",
+					);
+
+					let sliceMachineConfigExists = false;
+					try {
+						await this.manager.project.getSliceMachineConfigPath();
+						sliceMachineConfigExists = true;
+					} catch {
+						// noop, config does not exists, we'll create it
+					}
+
+					if (sliceMachineConfigExists) {
+						parentTask.title = "Updating Slice Machine configuration...";
+
+						const config = await this.manager.project.getSliceMachineConfig();
+						await this.manager.project.writeSliceMachineConfig({
+							config: {
+								...config,
+								repositoryName: this.context.repository.domain,
+								adapter: this.context.framework.adapterName,
+							},
+						});
+						parentTask.title = "Updated Slice Machine configuration";
+					} else {
+						parentTask.title = "Creating Slice Machine configuration...";
+
+						const sliceMachineConfigPath =
+							await this.manager.project.suggestSliceMachineConfigPath();
+
+						await this.manager.project.writeSliceMachineConfig({
+							config: {
+								// TODO: Update _latest to a real value or make it optional
+								_latest: "legacy",
```

Comment:
This is something I would like to discuss, `_latest` is just bad and has been added ad-hoc for migrations, that are themselves ad-hoc, clutter the code and have to be kept forever. How can the new architecture + plugins versioning help us get rid of this?

Reply:
I agree. `_latest` is fragile and exposes state into the `sm.json` configuration file.

In my opinion, the best way to handle migrations is to run a set of checks at the start of `start-slicemachine` and run migrations as needed. The checks would look at the state of the project at that time and deduce if a migration is needed.

For example, say we want to change the location of some files from `.slicemachine` into a Slice's directory. At start up, we would have a migration that does the following:

1. Read `.slicemachine` and look for the files that would need to be moved.
2. If we detect those files, move them.

These steps would run every time Slice Machine is run. We no longer "version" our migrations and instead always run them as needed. This is nice because it means if a developer undoes a migration, we'll detect it again and fix it.

The detection steps need to be quick and cheap since they are run each time SM starts. Ideally, each migration is non-blocking with no dependencies. That allows us to run all migrations in parallel.

Plugins may need to support migrations of their own. Today, we don't have a plugin hook that runs explicitly at SM start up like `on:start`. We may want to provide a dedicated `migration` hook instead which would allow SM to manage when migrations run. Initially, we may want migrations to run immediately, but in the future, we may want to delay that migration until a user clicks a button.

---

Migrations have been left out of the manager since I think we'll want to rearchitect how they work. The above is what I think would work well, but let's discuss this further before we implement anything.